purchase_order_reorder_lines/purchase.py

```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
+#
+#
+#    Author: Alexandre Fayolle
+#    Copyright 2013 Camptocamp SA
+#    Author: Damien Crier
+#    Copyright 2015 Camptocamp SA
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+
+from openerp import models, api, fields
+
+
+class purchase_order_line(models.Model):
+    _inherit = 'purchase.order.line'
+
+    sequence = fields.Integer(default=10,
+                              help="Gives the sequence of this line when "
+                                   "displaying the purchase order.")
+    _order = 'order_id desc, sequence, id'
+
+
+class purchase_order(models.Model):
+    _inherit = 'purchase.order'
+
+    @api.model
+    def _prepare_inv_line(self, account_id, order_line):
+        res = super(purchase_order, self)._prepare_inv_line(
+            account_id,
+            order_line,
+            )
+        res['sequence'] = order_line.sequence
+        return res
+
+    @api.model
+    def _prepare_order_line_move(self, order, order_line,
+                                 picking_id, group_id):
+        res = super(purchase_order, self)._prepare_order_line_move(
+            order,
+            order_line,
+            picking_id,
+            group_id)
+        if res:
+            res[0]['sequence'] = order_line.sequence
+        return res
+
+
+class purchase_line_invoice(models.TransientModel):
+    _inherit = 'purchase.order.line_invoice'
+
+    @api.multi
+    def makeInvoices(self):
+        invoice_line_obj = self.env['account.invoice.line']
+        purchase_line_obj = self.env['purchase.order.line']
+        res = super(purchase_line_invoice, self).makeInvoices()
+        invoice_ids = eval(res['domain'])[0][-1]  # OMG :-(
```

Comment:
And instead of taking `[0]` I would be reassured if you loop over the domain to find the right one, just in case someone changed the domain up in the calls chain.